### ISSUE TITLE
ENH: Compression

### ISSUE THREAD

**Issue by MartinThoma** [py-pdf/pypdf#1910]
## Explanation

PDF file size matters, especially if you need to send documents via e-mail or for people with bad internet. Smaller file sizes make handling the document easier.

Evidence that people care about this topic:

* [Optimize PDF files](https://stackoverflow.com/q/10450120/562769)
* [How can I lossless reduce PDF file size?](https://stackoverflow.com/q/58988258/562769)
* [How can I reduce the file size of PDF files in C#?](https://stackoverflow.com/q/33779412/562769)

What this is NOT about:

* Image size reductions done by the user: We describe that in https://pypdf.readthedocs.io/en/latest/user/file-size.html#removing-images .In the context of this enhancement proposal, we can also look at image formats / possibilities for size reduction, but it should require little user knowledge. For example, I could imagine that we automatically apply better compression algorithms.

## Goals

* Simpler file size reduction for users
* Better reduction

## Code Example

Similar to what we have already:

```python
from pypdf import PdfReader, PdfWriter

reader = PdfReader("example.pdf")
writer = PdfWriter()

for page in reader.pages:
    writer.add_page(page)

# That is the new thing:
# - some flag to indicate if lossy / lossless compression is desired
# - maybe a flag that steers CPU usage? Sometimes compression is rather time-intensive
writer.compress(image_quality=85)

with open("out.pdf", "wb") as f:
    writer.write(f)
```

**Comment by MartinThoma** [comment]
I've just seen a pretty crazy reduction with https://smallpdf.com/de/pdf-verkleinern :

* Original file: https://github.com/py-pdf/sample-files/blob/main/009-pdflatex-geotopo/GeoTopo.pdf (5.3 MB)
* Reduced size: [GeoTopo-komprimiert.pdf](https://github.com/py-pdf/pypdf/files/11845099/GeoTopo-komprimiert.pdf) (1.8 MB)

**Comment by pubpub-zz** [comment]
@MartinThoma should we keep it open ?

**Comment by pubpub-zz** [comment]
@mnmtz 
can you provide 2 samples for analysis.

**Comment by MartinThoma** [comment]
> should we keep it open ?

I want to apply everything from https://pypdf.readthedocs.io/en/latest/user/file-size.html to https://github.com/py-pdf/pypdf/issues/1910#issuecomment-1603764302 (GeoTopo.pdf) and see how far we get.

Let's say if we don't have concrete steps what to do here until 13th of August, we can close it.

**Comment by pubpub-zz** [comment]
@mnmtz can you provide the different ouput files to ease analysis

**Comment by MartinThoma** [comment]
Example 1 (pikepdf left, pypdf right):

![image](https://github.com/py-pdf/pypdf/assets/1658117/001acd85-d7a1-4389-a7ff-045dceebd4d6)

**Comment by pubpub-zz** [comment]
I've done a quick review with Pdf Box and I found that the contents shows very odd floating number:
```
(beginning of content on page 2)
/OC /MC0 BDC
EMC
/OC /MC1 BDC
1 0.882 0.339 0.251 k
/GS0 gs
0 0 613.27499999999998 128.452 re
f
BT
0.023 0.008999999999999999 0.005 0 k
/T1_0 1 Tf
0.003 Tc
-0.003 Tw
11 0 0 11 448.49509999999998 70.803700000000006 Tm
[ (\035\022) -4.5999999999999996 (\023) -26.5 <0716> -2.8999999999999999 (\015) -0.8 (\022) 1 (\035) -4 (\006) -8.8000000000000007 (\025) -4.7000000000000002 <16> -0.6 (\034) 5.5 (\014) -1 (\024) -4.7000000000000002 (\005) 12.6 (\010) 13 (\015\030) -3.2999999999999998 (\004) ] TJ
0 -1.5449999999999999 Td
[ (\035\022) -4.5999999999999996 (\023) -26.5 <0716> -2.8999999999999999 (\015) -0.8 (\022) 1 (\035) -1.2 (\010) 13.5 (\025) -4.7000000000000002 <161c> 5.5 (\014) -1 (\024) -4.7999999999999998 (\005) 12.6 (\010) 13 (\015\030) -3.2999999999999998 (\004) ] TJ
```

I've checked `FloatObject('-4.6')` which returns -4.5999999999999996

why ???
@MartinThoma  any ideas ?

**Comment by MartinThoma** [comment]
The issue is IEEE 754 double floating point precision. You can see that with https://baseconvert.com/ieee-754-floating-point

Python has some clever rounding logic for string representing that:

```
>>> -4.6
-4.6
```

The output here is actually not the true representation. It's rounding, but I don't know how Python does that.

**Comment by MartinThoma** [comment]
I think the shell uses the `g` floating point formatter: https://docs.python.org/3/library/string.html#format-specification-mini-language

> General format. For a given precision p >= 1, this rounds the number to p significant digits and then formats the result in either fixed-point format or in scientific notation, depending on its magnitude. A precision of 0 is treated as equivalent to a precision of 1.
>
>The precise rules are as follows: suppose that the result formatted with presentation type 'e' and precision p-1 would have exponent exp. Then, if m <= exp < p, where m is -4 for floats and -6 for [Decimals](https://docs.python.org/3/library/decimal.html#decimal.Decimal), the number is formatted with presentation type 'f' and precision p-1-exp. Otherwise, the number is formatted with presentation type 'e' and precision p-1. In both cases insignificant trailing zeros are removed from the significand, and the decimal point is also removed if there are no remaining digits following it, unless the '#' option is used.
>
> With no precision given, uses a precision of 6 significant digits for [float](https://docs.python.org/3/library/functions.html#float). For [Decimal](https://docs.python.org/3/library/decimal.html#decimal.Decimal), the coefficient of the result is formed from the coefficient digits of the value; scientific notation is used for values smaller than 1e-6 in absolute value and values where the place value of the least significant digit is larger than 1, and fixed-point notation is used otherwise.
>
> Positive and negative infinity, positive and negative zero, and nans, are formatted as inf, -inf, 0, -0 and nan respectively, regardless of the precision.

**Comment by pubpub-zz** [comment]
for me it is more complex:
```
float("-4.6")    # ->  returns -4.6

float.__new__(FloatObject,-4.6)  # -> returns -4.5999999999999996
```
I agree it seems to be linked with IEEE but I do not understand why it is not good.

**Comment by MartinThoma** [comment]
```
>>> from pypdf.generic import FloatObject
>>> float.__new__(FloatObject,-4.6) - (-4.6)
0.0
```

That was expected.

However, this is not expected:

```
>>> a = float.__new__(FloatObject,-4.6)
>>> a.__repr__ = float.__repr__
>>> a.__str__ = float.__str__
>>> a
-4.5999999999999996
```

**Comment by MartinThoma** [comment]
Ah, found it:

```
>>> FloatObject.__repr__ = float.__repr__
>>> a
-4.6
```

**Comment by MartinThoma** [comment]
Also:

```
>>> FloatObject.__repr__ = lambda self: f"{self:g}"
>>> a
-4.6
```

**Comment by johnkw** [comment]
What was the resolution to this bug? I only see a fix for float precision, but there seem to be much broader problems, or at least there were.  To compress PDFs I have to run pypdf output through a pipe of `gs -sDEVICE=pdfwrite` and it fixes things to be like 1/4 the size. In a good PDF there's very little text.

**Comment by stefan6419846** [comment]
In this specific case, the fix has indeed been to limit the float precision.

> but there seem to be much broader problems

Please be more specific in a dedicated issue.

> through a pipe of `gs -sDEVICE=pdfwrite`

This does all sorts of changes, including changes to the image resolution.

> In a good PDF there's very little text.

Please elaborate on this.

**Comment by johnkw** [comment]
I'm never using PDF images with pypdf.

Just loading any random PDF and re-writing with pypdf causes it to bloat in size with much more text instead of compressed binary. For example this adds a few 100k in size:
```
# random example file from https://www.primasoft.com/librarymanager/library-files/library-barcode-labels-20-avery-5161-2000.pdf
import pypdf
pdfwrite = pypdf.PdfWriter()
for page in pypdf.PdfReader(open('library-barcode-labels-20-avery-5161-2000.pdf','rb')).pages:
    p = pdfwrite.add_page(page)
    p.compress_content_streams(level=9) # this does close to nothing
pdfwrite.write(open('pypdf.pdf','wb'))
```

Piping through `gs -sDEVICE=pdfwrite` shrinks the pypdf output from 1324265 down to 939691.

**Comment by pubpub-zz** [comment]
Without the pdfs no analysis is possible

**Comment by johnkw** [comment]
I linked to a PDF in the example above. I just picked a random one from the internet: https://www.primasoft.com/librarymanager/library-files/library-barcode-labels-20-avery-5161-2000.pdf The particulars are irrelevant though. All PDFs I've ever looked at have this problem with pypdf. There is nothing unusual about the one I selected -- it was just the first that came up in a Google search for "example pdf" or something.
